### PR TITLE
Added support for train/val/test splits throughout project

### PR DESCRIPTION
**Pull request by jakeval**
* Each dataset has a single canonical train / validation / test split which is deterministically constructed after loading the data from memory.
* The models are trained and tuned using the train and validation set.
* The recourse methods are "fit" with the train set and can be tuned with the validation set.
* Different dataset splits are acquired by passing in a string or list of strings to load_data().

# Changes Made
* **data/data_loader.py** and **data/datasets/base_loader.py**: function signatures changed to allow a `split` argument indicating which split(s) to return.
* **data/datasets/credit_card_default_loader.py**: process_data() now deterministically splits the dataset into train/val/test sets.
* **run_experiment.py** now uses the training split to fit the recourse method and the config-determined split to run the recourse. Some error messages were also updated.
* **logistic_regression.py** and **model_trainer.py**: model training now uses the train/val/test splits. They also evaluate the confusion matrix instead of label-dependent accuracy.
* **generate_face_graph.py**: face graphs now use the training set.

# Feedback Requested
* Correctness -- are the splits used correctly throughout the project?
* Readability -- is it clear how the splits are generated and used?

### CODE REVIEW

File: models/core/model_trainer.py [hunk 1]
```diff
@@ -91,113 +94,80 @@ def new_model(
...
+            y_pred = model.predict(data)
+            y_true = data[dataset_info.label_column]
+
+            results = {}
+            results["accuracy"] = float((y_pred == y_true).mean())
+            results["true_positives"] = int(
+                (y_pred[pos_mask] == y_true[pos_mask]).sum()
+            )
+            results["false_positives"] = int(
+                (y_pred[pos_mask] != y_true[pos_mask]).sum()
```
Comment (bringingjoy) [inline models/core/model_trainer.py:156]: I think this should be `y_pred[~pos_mask] != y_true[~pos_mask]`.

File: models/core/model_trainer.py [hunk 2]
```diff
@@ -91,113 +94,80 @@ def new_model(
...
+            )
+            results["true_negatives"] = int(
+                (y_pred[~pos_mask] == y_true[~pos_mask]).sum()
+            )
+            results["false_negatives"] = int(
+                (y_pred[~pos_mask] != y_true[~pos_mask]).sum()
```
Comment (bringingjoy) [inline models/core/model_trainer.py:162]: I think this should be `y_pred[pos_mask] != y_true[pos_mask]`.

File: data/datasets/credit_card_default_loader.py [hunk 3]
```diff
@@ -98,4 +116,74 @@ def process_data(self, data: pd.DataFrame) -> pd.DataFrame:
...
+    ) -> Mapping[str, np.ndarray]:
+        """Partitions a set of indices given the train and validation ratios.
+
+        Returns a mapping from split name ('train', 'val', 'test') to the set
+        of indices that should appear in that split.
+
+        The dataset is assumed to be 0-indexed in increments of 1."""
+        if train_ratio < 0 or val_ratio < 0 or test_ratio < 0:
+            raise RuntimeError("Dataset split ratios must be non-negative.")
+        if np.abs((train_ratio + val_ratio + test_ratio) - 1) > 1e-5:
```
Comment (bringingjoy) [inline data/datasets/credit_card_default_loader.py:178]: Can you please add a comment to make it clear that this is checking that the different ratios sum to approx. 1?

**Review by bringingjoy — APPROVED** [review]
Looks good! A couple things to fix and then feel free to submit.

File: data/data_loader.py [hunk 4]
```diff
@@ -41,7 +45,7 @@ def load_data(
     elif dataset_name == DatasetName.TOY_DATASET:
         from confidence_checks import toy_data
```
[the remark this answers is not in the document]
Comment (jakeval) [reply data/data_loader.py:46]: I added [Issue 68](https://github.com/jakeval/MRMC/issues/68) addressing this.
